runtime/strategies/resolve-recipe.js

```diff
@@ -0,0 +1,143 @@
+        if (handle.connections.length == 0 || handle.id || (!handle.type) || (!handle.fate))
+          return;
+
+        const counts = RecipeUtil.directionCounts(handle);
+
+        let mappable;
+
+        switch (handle.fate) {
+          case 'use':
+            mappable = arc.findHandlesByType(handle.type, {tags: handle.tags, subtype: counts.out == 0});
+            break;
+          case 'map':
+          case 'copy':
+            mappable = arc.context.findHandlesByType(handle.type, {tags: handle.tags, subtype: true});
+            break;
+          case 'create':
+          case '?':
+            mappable = [];
+            break;
+          default:
+            assert(false, `unexpected fate ${handle.fate}`);
+        }
+
+        mappable = mappable.filter(incomingHandle => {
+          for (let existingHandle of recipe.handles)
+            if (incomingHandle.id == existingHandle.id)
+              return false;
+          return true;
+        });
+
+        if (mappable.length == 1) {
+          return (recipe, handle) => {
+            handle.mapToView(mappable[0]);
+          };            
+        }
+      }
+
+      onSlotConnection(recipe, slotConnection) {
+        if (this._slotIsConnected(slotConnection))
+          return;
+        
+        let selectedSlots = [];
+        if (!slotConnection.targetSlot)
+          selectedSlots = this._findSlotCandidates(slotConnection, recipe.slots);
+        
+        selectedSlots = selectedSlots.concat(this._findSlotCandidates(slotConnection, arc.pec.slotComposer.getAvailableSlots()));
+      
+        if (selectedSlots.length !== 1)
+          return;
+        
+        let selectedSlot = selectedSlots[0];
+
+        return (recipe, slotConnection) => {
+          if (!slotConnection.targetSlot) {
+            let clonedSlot = recipe.updateToClone({selectedSlot}).selectedSlot;
+
+            if (!clonedSlot) {
+              clonedSlot = recipe.slots.find(s => selectedSlot.id && selectedSlot.id == s.id);
+              if (clonedSlot == undefined)
+                clonedSlot = recipe.newSlot(selectedSlot.name);
+                clonedSlot.id = selectedSlot.id;
+            }
+            slotConnection.connectToSlot(clonedSlot);
+          }
+
+          assert(!selectedSlot.id || !slotConnection.targetSlot.id || (selectedSlot.id == slotConnection.targetSlot.id),
+            `Cannot override slot id '${slotConnection.targetSlot.id}' with '${selectedSlot.id}'`);
+          slotConnection.targetSlot.id = selectedSlot.id || slotConnection.targetSlot.id;
+
+          // TODO: need to concat to existing tags and dedup?
+          slotConnection.targetSlot.tags = [...selectedSlot.tags];
+          return 1;
+        };
+      }
+
+      _slotIsConnected(slotConnection) {
+        return slotConnection.targetSlot && ((!!slotConnection.targetSlot.sourceConnection) || (!!slotConnection.targetSlot.id));
+      }
+
+      _findSlotCandidates(slotConnection, slots) {
```

Comment:
For now, I think we could make these functions on MapSlots be `static` (I don't think they use `this` other than to call each other)? I think much of this should ultimately be moved to recipe/

Reply:
Done